File: stream-ar/Assets/Shared/GraphLinks/Scripts/Link.cs
using Assets.Modules.Core;
using Assets.Modules.Interactions;
using Assets.Modules.Networking;
using Newtonsoft.Json.Linq;
using System.Linq;
using UniRx;
using UnityEngine;

namespace Assets.Modules.GraphLinks
{
    [SelectionBase]
    public class Link : ObservableModel<Link>, IInteractable
    {
        const float LINK_USER_PROXIMITY = 1.1f;

        public string GetInteractionType() => "link";
        public int GetInteractionId() => Id;


        public Transform Visualization;

        public readonly ReactiveProperty<int> UpstreamRx = new ReactiveProperty<int>(-1);
        public int Upstream
        {
            get => UpstreamRx.Value;
            set
            {
                if (UpstreamRx.Value != value)
                {
                    UpstreamRx.Value = value;
                    TriggerLocalChange("Upstream");
                }
            }
        }



        public readonly ReactiveProperty<int> DownstreamRx = new ReactiveProperty<int>(-1);
        public int Downstream
        {
            get => DownstreamRx.Value;
            set
            {
                if (DownstreamRx.Value != value)
                {
                    DownstreamRx.Value = value;
                    TriggerLocalChange("Downstream");
                }
            }
        }



        private Vector3 _placingPosition = Vector3.zero;
        public Vector3 PlacingPosition
        {
            get => _placingPosition;
            set
            {
                if (_placingPosition != value)
                {
                    _placingPosition = value;
                    TriggerLocalChange("PlacingPosition");
                }
            }
        }



        private Quaternion _placingRotation = Quaternion.identity;
        public Quaternion PlacingRotation
        {
            get => _placingRotation;
            set
            {
                if (_placingRotation != value)
                {
                    _placingRotation = value;
                    TriggerLocalChange("PlacingRotation");
                }
            }
        }


        public readonly ReactiveProperty<int> CreatedByRx = new ReactiveProperty<int>(-1);
        public int CreatedBy
        {
            get => CreatedByRx.Value;
            set
            {
                if (CreatedByRx.Value != value)
                {
                    CreatedByRx.Value = value;
                    TriggerLocalChange("CreatedBy");
                }
            }
        }


        public readonly ReactiveProperty<bool> UseSortRx = new ReactiveProperty<bool>(false);
        public bool UseSort
        {
            get => UseSortRx.Value;
            set
            {
                if (UseSortRx.Value != value)
                {
                    UseSortRx.Value = value;
                    TriggerLocalChange("UseSort");
                }
            }
        }


        public readonly ReactiveProperty<bool> UseColorRx = new ReactiveProperty<bool>(false);
        public bool UseColor
        {
            get => UseColorRx.Value;
            set
            {
                if (UseColorRx.Value != value)
                {
                    UseColorRx.Value = value;
                    TriggerLocalChange("UseColor");
                }
            }
        }


#if STREAM_AR

        public ReactiveProperty<bool> IsUserInProximity { get; private set; } = new ReactiveProperty<bool>(false);


        private StreamClient _client;
        private Transform _cam;
        private void OnEnable()
        {
            _client = StreamClient.Instance;
            _cam = Camera.main.transform;
        }


        private void Update()
        {
            IsUserInProximity.Value = (_client.LookingAtType == "link" && _client.LookingAtId == Id)
                || (_client.SelectedType == "link" && _client.SelectedId == Id)
                || Mathf.Abs((_cam.position - Visualization.position).magnitude) < LINK_USER_PROXIMITY;
        }

#endif


        protected override void OnDestroy()
        {
            base.OnDestroy();
            CreatedByRx.Dispose();
            DownstreamRx.Dispose();
            UpstreamRx.Dispose();

#if STREAM_AR
            IsUserInProximity.Dispose();
#endif
        }

        public override JObject ToJson()
        {
            return new JObject
            {
                { "id", Id },
                { "upstream", Upstream },
                { "downstream", Downstream },
                { "useColor", UseColor },
                { "useSort", UseSort },
                { "createdBy", CreatedBy },
                { "placingPosition", new JArray { _placingPosition.x, _placingPosition.y, _placingPosition.z } },
                { "placingRotation", new JArray { _placingRotation.x, _placingRotation.y, _placingRotation.z, _placingRotation.w } },
            };
        }

        protected override void ApplyRemoteUpdate(JObject updates)
        {
            var upstream = updates["upstream"];
            if (upstream != null)
                UpstreamRx.Value = upstream.Value<int>();

            var downstream = updates["downstream"];
            if (downstream != null)
                DownstreamRx.Value = downstream.Value<int>();

            var jPos = updates["placingPosition"];
            if (jPos != null)
            {
                var pos = jPos.Select(m => (float)m).ToArray();
                if (pos.Length == 3)
                    _placingPosition = new Vector3(pos[0], pos[1], pos[2]);
            }

            var jRot = updates["placingRotation"];
            if (jRot != null)
            {
                var rot = jRot.Select(m => (float)m).ToArray();
                if (rot.Length == 4)
                    _placingRotation = new Quaternion(rot[0], rot[1], rot[2], rot[3]);
            }


            var createdBy = updates["createdBy"];
            if (createdBy != null)
                CreatedByRx.Value = createdBy.Value<int>();


            var useColor = updates["useColor"];
            if (useColor != null)
                UseColorRx.Value = useColor.Value<bool>();

            var useSort = updates["useSort"];
            if (useSort != null)
                UseSortRx.Value = useSort.Value<bool>();
        }
    }
}

File: stream-tablet/src/app/models/link.ts
import * as _ from 'lodash';
import { ObservableModel } from './observable-model';

export class Link extends ObservableModel {

    private _upstream = -1;
    public get upstream() {
        return this._upstream;
    }

    public set upstream(val: number) {
        if (this._upstream !== val) {
            this._upstream = val;
            this.modelChanges.next('upstream');
        }
    }

    private _downstream = -1;
    public get downstream() {
        return this._downstream;
    }

    public set downstream(val: number) {
        if (this._downstream !== val) {
            this._downstream = val;
            this.modelChanges.next('downstream');
        }
    }

    // only temporary, if _up or _downstream == -1
    private _placingPosition: [number, number, number] = [0, 0, 0];
    public get placingPosition() {
        return this._placingPosition;
    }

    public set placingPosition(val: [number, number, number]) {
        if (!_.isEqual(this._placingPosition, val)) {
            this._placingPosition = val;
            this.modelChanges.next('placingPosition');
        }
    }


    private _placingRotation: [number, number, number, number] = [0, 0, 0, 1];
    public get placingRotation() {
        return this._placingRotation;
    }

    public set placingRotation(val: [number, number, number, number]) {
        if (!_.isEqual(this._placingRotation, val)) {
            this._placingRotation = val;
            this.modelChanges.next('placingRotation');
        }
    }


    private _createdBy = -1;
    public get createdBy() {
        return this._createdBy;
    }

    public set createdBy(val: number) {
        if (this._createdBy !== val) {
            this._createdBy = val;
            this.modelChanges.next('createdBy');
        }
    }


    private _useColor = false;
    public get useColor() {
        return this._useColor;
    }

    public set useColor(val: boolean) {
        if (this._useColor !== val) {
            this._useColor = val;
            this.modelChanges.next('useColor');
        }
    }


    private _useSort = false;
    public get useSort() {
        return this._useSort;
    }

    public set useSort(val: boolean) {
        if (this._useSort !== val) {
            this._useSort = val;
            this.modelChanges.next('useSort');
        }
    }
}
